chainer/functions/math/tensordot.py

```diff
@@ -0,0 +1,171 @@
+import collections
+import numpy
+import six
+
+from chainer.backends import cuda
+from chainer import function_node
+from chainer import utils
+from chainer.utils import type_check
+
+
+def _tensordot(a, b, a_axes, b_axes, c_axes):
+
+    a_row_ndim = len(a_axes[0])
+    a_col_ndim = len(a_axes[1])
+    b_row_ndim = len(b_axes[0])
+    b_col_ndim = len(b_axes[1])
+    c_row_ndim = len(c_axes[0])
+    c_col_ndim = len(c_axes[1])
+    if a_col_ndim != b_row_ndim:
+        raise ValueError('axes count mismatch')
+    if a_row_ndim != c_row_ndim:
+        raise ValueError('axes count mismatch')
+    if b_col_ndim != c_col_ndim:
+        raise ValueError('axes count mismatch')
+    if a.ndim < a_col_ndim or b.ndim < b_row_ndim:
+        raise ValueError('dimension of input tensors must be '
+                         'greater equal to dot-axes count ({})'
+                         .format(a_col_ndim))
+    for a_axis, b_axis in zip(a_axes[1], b_axes[0]):
+        if a.shape[a_axis] != b.shape[b_axis]:
+            raise ValueError('shape mismatch')
+
+    xp = cuda.get_array_module(a)
+    y = xp.tensordot(a, b, axes=(a_axes[1], b_axes[0]))
+
+    trans = [None for i in six.moves.range(y.ndim)]
+    table_a = [1 if i in a_axes[0] else 0 for i in six.moves.range(a.ndim)]
+    table_a = numpy.cumsum(table_a) - 1
+    for i, c_axis in enumerate(c_axes[0]):
+        trans[c_axis] = table_a[a_axes[0][i]]
+    table_b = [1 if i in b_axes[1] else 0 for i in six.moves.range(b.ndim)]
+    table_b = numpy.cumsum(table_b) - 1
+    for i, c_axis in enumerate(c_axes[1]):
+        trans[c_axis] = table_b[b_axes[1][i]] + len(a_axes[0])
+    for i, c_axis in enumerate(trans):
+        if i != c_axis:
+            y = xp.transpose(y, trans)
+            break
+
+    return y
+
+
+class TensorDot(function_node.FunctionNode):
+
+    def __init__(self, axes=2, a_axes=None, b_axes=None, c_axes=None,
+                 dtype=None):
+        self.axes = axes
+        self.a_axes = a_axes
+        self.b_axes = b_axes
+        self.c_axes = c_axes
+        self.dtype = dtype
+
+        if isinstance(axes, collections.Sequence):
+            if len(axes) != 2:
+                raise ValueError('axes must consist of two arrays.')
+        elif isinstance(axes, int):
+            pass
+        else:
+            raise TypeError('axes must be int or tuple of int')
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        a_type, b_type = in_types
+
+        type_check.expect(
+            a_type.dtype.kind == 'f',
+            b_type.dtype.kind == 'f',
+            a_type.ndim >= 1,
+            b_type.ndim >= 1,
```

Comment:
Is ndim of 0 too trivial? Since numpy/cupy support it, I think I'd be nice if Chainer could too.

Reply:
Agree. will check if it works well when ndim is zero.

Reply:
I'm wondering how to handle an input with ndim=0, mainly because `chainer.Variable` does not accepts a scalar as its input. If you make it ndarray with shape of (1), then it becomes ndim=1.
Well, do we really need to consider the case of ndim=0?

Reply:
Ah, I think you can create a non-scalar ndarray with ndim=0 using e.g. 

```
x_data = numpy.random.uniform(-1, 1, ()))
x = chainer.Variable(x)
```

I am not sure how important this feature is but I noticed that e.g. matmul does support this case. Does it require a lot of changes to the code? https://github.com/chainer/chainer/blob/master/chainer/functions/math/matmul.py#L83

Reply:
Thank you!  I didn't know how to make a ndarray with ndim=0. I don't think it requires a lot of changes to support it. will fix soon.